ROS2/ROS_workspace/cli_app/src/gui.cpp

```diff
@@ -0,0 +1,414 @@
+#if defined(IMGUI_IMPL_OPENGL_ES2)
+#include <GLES2/gl2.h>
+#endif
+#include <GLFW/glfw3.h> // Will drag system OpenGL headers
+#include <opencv2/opencv.hpp>
+#include "rclcpp/rclcpp.hpp"
+#include "eer_messages/msg/pilot_input.hpp"
+#include "controller.h"
+
+// [Win32] Our example includes a copy of glfw3.lib pre-compiled with VS2010 to maximize ease of testing and compatibility with old VS compilers.
+// To link with VS2010-era libraries, VS2015+ requires linking with legacy_stdio_definitions.lib, which we do using this pragma.
+// Your own project should not be affected, as you are likely to link with a newer binary of GLFW that is adequate for your version of Visual Studio.
+#if defined(_MSC_VER) && (_MSC_VER >= 1900) && !defined(IMGUI_DISABLE_WIN32_FUNCTIONS)
+#pragma comment(lib, "legacy_stdio_definitions")
+#endif
+
+// This example can also compile and run with Emscripten! See 'Makefile.emscripten' for details.
+#ifdef __EMSCRIPTEN__
+#include "../libs/emscripten/emscripten_mainloop_stub.h"
+#endif
+
+static void glfw_error_callback(int error, const char* description)
+{
+    fprintf(stderr, "GLFW Error %d: %s\n", error, description);
+}
+
+void webcamCapture(std::atomic<bool>& running, std::atomic<bool>& show_video, cv::Mat& frame, std::mutex& frame_mutex) {
+    cv::VideoCapture cap(0);
+    if (!cap.isOpened()) {
+        std::cerr << "Failed to open webcam" << std::endl;
+        running.store(false);
+        return;
+    }
+
+    // Set capture properties for better performance
+    cap.set(cv::CAP_PROP_FRAME_WIDTH, 320); // Lower resolution
+    cap.set(cv::CAP_PROP_FRAME_HEIGHT, 240); // Lower resolution
+    cap.set(cv::CAP_PROP_FPS, 30); // Limit frame rate
+
+    while (running.load()) {
+        if (show_video.load()) {
+            cv::Mat temp_frame;
+            cap >> temp_frame;
+            if (!temp_frame.empty()) {
+                std::lock_guard<std::mutex> lock(frame_mutex);
+                frame = temp_frame.clone();
+            }
+        } else {
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        }
+    }
+
+    cap.release();
+}
+
+
+GLuint matToTexture(const cv::Mat &mat, GLuint textureID, GLenum minFilter, GLenum magFilter, GLenum wrapFilter)
+{
+    if (textureID == 0) {
+        glGenTextures(1, &textureID);
+        glBindTexture(GL_TEXTURE_2D, textureID);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, minFilter);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, magFilter);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, wrapFilter);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, wrapFilter);
+        glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, mat.cols, mat.rows, 0, GL_BGR, GL_UNSIGNED_BYTE, mat.ptr());
+    } else {
+        glBindTexture(GL_TEXTURE_2D, textureID);
+        glTexSubImage2D(GL_TEXTURE_2D, 0, 0, 0, mat.cols, mat.rows, GL_BGR, GL_UNSIGNED_BYTE, mat.ptr());
+    }
+
+    return textureID;
+}
```

Comment:
Might be worth adding a comment like "This function converts an OpenCV cv::Mat image to an OpenGL texture"